File: packages/channels/joyspace/src/export.ts
import fs from "node:fs";
import path from "node:path";
import type { JacpCredentials, Logger } from "./types.js";
import {
  createJoySpaceApiContext,
  downloadDiagramXml,
  extractPageIdFromUrl,
  fetchDiagramDetail,
  fetchPageBasic,
  fetchPageContent,
} from "./vendor/joyspace-api-client.mjs";
import { joyspaceContentToMarkdown } from "./vendor/joyspace-content-to-markdown.mjs";
import { drawioXmlToMermaid } from "./vendor/joyspace-drawio-to-mermaid.mjs";

export function isJoySpaceFixtureMode(): boolean {
  return process.env.PORIA_JOYSPACE_FIXTURE === "1";
}

export type ExportJoySpaceInput = {
  url: string;
  outputDir?: string;
  outputName?: string;
  tenantCode?: string;
  credentials?: JacpCredentials;
  logger?: Logger;
};

export type ExportJoySpaceResult = {
  outputPath: string;
  title: string;
  cookieSource?: string;
};

function sanitizeFilename(value: string): string {
  const cleaned = String(value || "Doc")
    .replace(/[/\\:*?"<>|]/g, "_")
    .replace(/\s+/g, " ")
    .trim();
  return cleaned || "Doc";
}

function ensureMarkdownSuffix(filename: string): string {
  return filename.toLowerCase().endsWith(".md") ? filename : `${filename}.md`;
}

function resolveOutputPath(outputDir: string, filename: string): string {
  const initialPath = path.join(outputDir, ensureMarkdownSuffix(filename));
  if (!fs.existsSync(initialPath)) {
    return initialPath;
  }

  const parsed = path.parse(initialPath);
  let index = 1;
  while (true) {
    const candidate = path.join(
      parsed.dir,
      `${parsed.name}-${index}${parsed.ext}`,
    );
    if (!fs.existsSync(candidate)) {
      return candidate;
    }
    index += 1;
  }
}

function collectDiagramIds(content: unknown[]): string[] {
  const ids = new Set<string>();
  const visit = (nodes: unknown): void => {
    if (!Array.isArray(nodes)) return;
    for (const node of nodes) {
      if (!node || typeof node !== "object") continue;
      const record = node as Record<string, unknown>;
      if (record.type === "diagram") {
        const id = String(record.diagramId || record.id || "").trim();
        if (id) ids.add(id);
      }
      if (Array.isArray(record.children)) {
        visit(record.children);
      }
    }
  };
  visit(content);
  return [...ids];
}

async function renderSvgFromXml(xml: string, logger?: Logger): Promise<string> {
  const exportUrl = process.env.JOYSPACE_DRAWIO_EXPORT_URL;
  if (!exportUrl || !xml) return "";
  try {
    const response = await fetch(exportUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/xml",
        Accept: "image/svg+xml",
      },
      body: xml,
    });
    if (!response.ok) {
      logger?.warn?.(
        `[diagram] SVG render failed HTTP ${response.status} ${response.statusText}`,
      );
      return "";
    }
    const svg = await response.text();
    return svg.includes("<svg") ? svg : "";
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger?.warn?.(`[diagram] SVG render error: ${message}`);
    return "";
  }
}

type DiagramInfo = {
  title: string;
  linkUrl: string;
  pageUrl: string;
  svg: string;
  mermaid: string;
  error?: string;
};

type ApiContext = Awaited<ReturnType<typeof createJoySpaceApiContext>>;

async function fetchDiagramEntry(input: {
  diagramId: string;
  pageId: string;
  apiContext: ApiContext;
  pageUrl: string;
  logger?: Logger;
}): Promise<[string, DiagramInfo]> {
  const { diagramId, pageId, apiContext, pageUrl, logger } = input;
  try {
    const detail = await fetchDiagramDetail({
      diagramId,
      pageId,
      cookieHeader: apiContext.cookieHeader,
      teamHeaderId: apiContext.teamHeaderId,
    });

    const linkUrl = String(
      detail?.linkUrl || detail?.link_url || detail?.url || "",
    );
    const title = String(
      detail?.title || detail?.name || detail?.fileName || "",
    ).trim();

    let xml = "";
    if (linkUrl) {
      try {
        xml = await downloadDiagramXml(linkUrl);
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        logger?.warn?.(`[diagram ${diagramId}] XML download failed: ${message}`);
      }
    }

    const svg = xml ? await renderSvgFromXml(xml, logger) : "";
    let mermaid = "";
    if (xml && !svg) {
      try {
        mermaid = drawioXmlToMermaid(xml) || "";
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        logger?.warn?.(
          `[diagram ${diagramId}] mermaid conversion failed: ${message}`,
        );
      }
    }

    return [
      diagramId,
      { title, linkUrl, pageUrl, svg, mermaid },
    ];
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    logger?.warn?.(`[diagram ${diagramId}] detail fetch failed: ${message}`);
    return [
      diagramId,
      {
        title: "",
        linkUrl: "",
        pageUrl,
        svg: "",
        mermaid: "",
        error: message,
      },
    ];
  }
}

async function fetchDiagramsMap(input: {
  content: unknown[];
  pageId: string;
  apiContext: ApiContext;
  pageUrl: string;
  logger?: Logger;
  concurrency?: number;
}): Promise<Map<string, DiagramInfo>> {
  const {
    content,
    pageId,
    apiContext,
    pageUrl,
    logger,
    concurrency = 4,
  } = input;
  const ids = collectDiagramIds(content);
  const diagrams = new Map<string, DiagramInfo>();
  if (ids.length === 0) return diagrams;

  let cursor = 0;
  const runWorker = async (): Promise<void> => {
    while (cursor < ids.length) {
      const index = cursor;
      cursor += 1;
      const diagramId = ids[index]!;
      const [, info] = await fetchDiagramEntry({
        diagramId,
        pageId,
        apiContext,
        pageUrl,
        logger,
      });
      diagrams.set(diagramId, info);
    }
  };

  await Promise.all(
    Array.from({ length: Math.min(concurrency, ids.length) }, () =>
      runWorker(),
    ),
  );
  return diagrams;
}

function writeFixtureMarkdown(input: ExportJoySpaceInput): ExportJoySpaceResult {
  const outputDir = path.resolve(input.outputDir || process.cwd());
  fs.mkdirSync(outputDir, { recursive: true });
  const title = "JoySpace Fixture Page";
  const requestedName = input.outputName
    ? sanitizeFilename(input.outputName)
    : sanitizeFilename(title);
  const outputPath = resolveOutputPath(outputDir, requestedName);
  const markdown = [
    `# ${title}`,
    "",
    `Source: ${input.url}`,
    "",
    "This is offline fixture markdown from `PORIA_JOYSPACE_FIXTURE=1`.",
    "",
  ].join("\n");
  fs.writeFileSync(outputPath, markdown, "utf8");
  return { outputPath, title, cookieSource: "fixture" };
}

export async function exportJoySpaceMarkdown(
  input: ExportJoySpaceInput,
): Promise<ExportJoySpaceResult> {
  if (isJoySpaceFixtureMode()) {
    return writeFixtureMarkdown(input);
  }

  const url = input.url?.trim();
  if (!url) {
    throw new Error("JoySpace URL is required");
  }

  const pageId = extractPageIdFromUrl(url);
  const outputDir = path.resolve(input.outputDir || process.cwd());
  fs.mkdirSync(outputDir, { recursive: true });

  const credentialOverrides = input.credentials?.cookie
    ? {
        cookieHeader: input.credentials.cookie,
        cookieSource: input.credentials.username
          ? `poria-auth:${input.credentials.username}`
          : "poria-auth",
      }
    : {};

  const apiContext = await createJoySpaceApiContext({
    tenantCode: input.tenantCode,
    ...credentialOverrides,
  });

  const basic = await fetchPageBasic({
    pageId,
    cookieHeader: apiContext.cookieHeader,
    teamHeaderId: apiContext.teamHeaderId,
  });

  let effectivePageId = pageId;
  let titleSource: Record<string, unknown> = basic;
  const pageType = basic?.page_type ?? basic?.type;
  let usedOriginId: string | null = null;

  if (pageType === 5 && basic?.origin_id) {
    effectivePageId = String(basic.origin_id);
    usedOriginId = effectivePageId;
    input.logger?.info?.(
      `[info] page_type=${pageType} (reference page), falling back to origin_id=${effectivePageId}`,
    );
    try {
      const originBasic = await fetchPageBasic({
        pageId: effectivePageId,
        cookieHeader: apiContext.cookieHeader,
        teamHeaderId: apiContext.teamHeaderId,
      });
      titleSource = originBasic;
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      input.logger?.warn?.(`[warn] Failed to fetch origin basic: ${message}`);
    }
  } else if (pageType && !["13", 13, 5, "5"].includes(String(pageType))) {
    input.logger?.warn?.(
      `[warn] page_type=${pageType} may not be fully supported`,
    );
  }

  const contentPayload = await fetchPageContent({
    pageId: effectivePageId,
    cookieHeader: apiContext.cookieHeader,
    teamHeaderId: apiContext.teamHeaderId,
  });

  const content = contentPayload?.content;
  if (!Array.isArray(content) || content.length === 0) {
    throw new Error(
      `JoySpace API returned no valid data.content. page_type=${pageType ?? "unknown"}, pageId=${pageId}${
        usedOriginId ? ` (used origin ${usedOriginId})` : ""
      }`,
    );
  }

  const diagrams = await fetchDiagramsMap({
    content,
    pageId: effectivePageId,
    apiContext,
    pageUrl: url,
    logger: input.logger,
  });

  const conversion = joyspaceContentToMarkdown({
    title: String(
      titleSource?.title ||
        titleSource?.full_name ||
        basic?.title ||
        basic?.full_name ||
        "Doc",
    ),
    content,
    diagrams,
    pageUrl: url,
  });

  if (!conversion.markdown.trim()) {
    throw new Error("Markdown conversion produced empty output");
  }

  const requestedName = input.outputName
    ? sanitizeFilename(input.outputName)
    : sanitizeFilename(conversion.title);
  const outputPath = resolveOutputPath(outputDir, requestedName);
  const markdown = conversion.markdown.endsWith("\n")
    ? conversion.markdown
    : `${conversion.markdown}\n`;

  fs.writeFileSync(outputPath, markdown, "utf8");

  return {
    outputPath,
    title: conversion.title,
    cookieSource: apiContext.cookieSource || undefined,
  };
}

File: crates/poria-channels/src/joyspace/export.rs
use std::collections::{HashMap, HashSet};

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::client::{
    extract_page_id_from_url, fetch_diagram_detail, fetch_page_basic, fetch_page_content,
    JoySpaceAuth,
};
use super::markdown::{joyspace_content_to_markdown, ConversionResult, DiagramInfo};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExportedDocument {
    pub title: String,
    pub markdown: String,
    pub page_id: String,
    pub effective_page_id: String,
    pub warnings: Vec<String>,
}

fn json_i64(value: &Value) -> Option<i64> {
    value
        .as_i64()
        .or_else(|| value.as_u64().map(|n| n as i64))
        .or_else(|| value.as_str().and_then(|s| s.parse().ok()))
}

fn json_str(value: &Value, keys: &[&str]) -> String {
    for key in keys {
        if let Some(text) = value.get(*key).and_then(Value::as_str) {
            let trimmed = text.trim();
            if !trimmed.is_empty() {
                return trimmed.to_string();
            }
        }
    }
    String::new()
}

fn collect_diagram_ids(nodes: &[Value]) -> Vec<String> {
    let mut ids = Vec::new();
    let mut seen = HashSet::new();
    fn visit(nodes: &[Value], ids: &mut Vec<String>, seen: &mut HashSet<String>) {
        for node in nodes {
            if node.get("type").and_then(Value::as_str) == Some("diagram") {
                let id = node
                    .get("diagramId")
                    .or_else(|| node.get("id"))
                    .and_then(Value::as_str)
                    .unwrap_or("")
                    .trim()
                    .to_string();
                if !id.is_empty() && seen.insert(id.clone()) {
                    ids.push(id);
                }
            }
            if let Some(children) = node.get("children").and_then(Value::as_array) {
                visit(children, ids, seen);
            }
        }
    }
    visit(nodes, &mut ids, &mut seen);
    ids
}

async fn fetch_diagrams(
    auth: &JoySpaceAuth,
    content: &[Value],
    page_id: &str,
    page_url: &str,
) -> HashMap<String, DiagramInfo> {
    let mut diagrams = HashMap::new();
    for diagram_id in collect_diagram_ids(content) {
        match fetch_diagram_detail(auth, &diagram_id, page_id).await {
            Ok(detail) => {
                let title = json_str(&detail, &["title", "name", "fileName"]);
                let link_url = json_str(&detail, &["linkUrl", "link_url", "url"]);
                diagrams.insert(
                    diagram_id,
                    DiagramInfo {
                        title,
                        link_url,
                        page_url: page_url.to_string(),
                        svg: String::new(),
                        mermaid: String::new(),
                    },
                );
            }
            Err(_) => {
                diagrams.insert(
                    diagram_id,
                    DiagramInfo {
                        page_url: page_url.to_string(),
                        ..Default::default()
                    },
                );
            }
        }
    }
    diagrams
}

fn conversion_to_export(
    conversion: ConversionResult,
    page_id: String,
    effective_page_id: String,
) -> Result<ExportedDocument, Box<dyn std::error::Error + Send + Sync>> {
    if conversion.markdown.trim().is_empty() {
        return Err("Markdown 转换结果为空。".into());
    }
    let markdown = if conversion.markdown.ends_with('\n') {
        conversion.markdown
    } else {
        format!("{}\n", conversion.markdown)
    };
    Ok(ExportedDocument {
        title: conversion.title,
        markdown,
        page_id,
        effective_page_id,
        warnings: conversion.warnings,
    })
}

/// Fetch a JoySpace page via SSO cookie + POST `/v1/pages/content` and convert to Markdown.
pub async fn export_page_markdown(
    auth: &JoySpaceAuth,
    url: &str,
) -> Result<ExportedDocument, Box<dyn std::error::Error + Send + Sync>> {
    let page_id = extract_page_id_from_url(url)?;
    let basic = fetch_page_basic(auth, &page_id).await?;

    let mut effective_page_id = page_id.clone();
    let mut title_source = basic.clone();
    let page_type = basic
        .get("page_type")
        .or_else(|| basic.get("type"))
        .and_then(json_i64);
    if page_type == Some(5) {
        if let Some(origin_id) = basic.get("origin_id").and_then(Value::as_str) {
            let origin_id = origin_id.trim();
            if !origin_id.is_empty() {
                effective_page_id = origin_id.to_string();
                if let Ok(origin_basic) = fetch_page_basic(auth, &effective_page_id).await {
                    title_source = origin_basic;
                }
            }
        }
    }

    let content_payload = fetch_page_content(auth, &effective_page_id).await?;
    let content = content_payload
        .get("content")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    if content.is_empty() {
        return Err(format!(
            "JoySpace API 未返回有效的 data.content。pageId={page_id}"
        )
        .into());
    }

    let diagrams = fetch_diagrams(auth, &content, &effective_page_id, url).await;
    let title = json_str(&title_source, &["title", "full_name"]);
    let title = if title.is_empty() {
        json_str(&basic, &["title", "full_name"])
    } else {
        title
    };
    let conversion = joyspace_content_to_markdown(&title, &content, &diagrams, url);
    conversion_to_export(conversion, page_id, effective_page_id)
}

/// Convert an already-fetched content array (used by tests / fixture).
pub fn markdown_from_content(
    title: &str,
    content: &[Value],
    page_url: &str,
) -> Result<ExportedDocument, Box<dyn std::error::Error + Send + Sync>> {
    let conversion = joyspace_content_to_markdown(title, content, &HashMap::new(), page_url);
    conversion_to_export(conversion, String::new(), String::new())
}
